cli/source/hooks/usePreCheck.ts

```diff
@@ -0,0 +1,80 @@
+import { isFreePort } from 'find-free-ports';
+import { useState } from 'react';
+
+import fs from 'fs';
+
+import { PreCheckStates } from '../constants.js';
+import { runCommand } from '../utils/run-command.js';
+
+export const usePreCheck = ({
+  db,
+  redis,
+  frontend,
+  sync_server,
+  analytics_server
+}: {
+  db: number;
+  redis: number;
+  frontend: number;
+  sync_server: number;
+  analytics_server: number;
+}) => {
+  const [daemon, setDaemon] = useState<PreCheckStates>(PreCheckStates.RUNNING);
+  const [ports, setPorts] = useState<PreCheckStates>(PreCheckStates.RUNNING);
+  const [composeFile, setComposeFile] = useState<PreCheckStates>(
+    PreCheckStates.RUNNING
+  );
+  const [dockerFile, setDockerFile] = useState<PreCheckStates>(
+    PreCheckStates.RUNNING
+  );
+
+  const callChecks = async () => {
+    // For Docker daemon
+    runCommand('docker', ['info'])
+      .promise.then(() => {
+        setDaemon(PreCheckStates.SUCCESS);
+      })
+      .catch((err) => {
+        setDaemon(PreCheckStates.FAILED);
+      });
+
+    // For ports
+    if (
+      isNaN(db) ||
+      isNaN(redis) ||
+      isNaN(frontend) ||
+      isNaN(sync_server) ||
+      isNaN(analytics_server)
+    ) {
+      setPorts(PreCheckStates.FAILED);
+    } else {
+      const db_check = await isFreePort(db);
+      const redis_check = await isFreePort(redis);
+      const fe_check = await isFreePort(frontend);
+      const ss_check = await isFreePort(sync_server);
+      const as_check = await isFreePort(analytics_server);
+      if (!db_check || !redis_check || !fe_check || !ss_check || !as_check) {
+        setPorts(PreCheckStates.FAILED);
+      } else {
+        setPorts(PreCheckStates.SUCCESS);
+      }
+    }
+
+    // For files
+    fs.promises
+      .access('../docker-compose.yml')
+      .then(() => {
+        setComposeFile(PreCheckStates.SUCCESS);
+      })
+      .catch((err) => {
+        setComposeFile(PreCheckStates.FAILED);
+      });
+
+    fs.promises
+      .access('../Dockerfile.dev')
+      .then(() => setDockerFile(PreCheckStates.SUCCESS))
+      .catch(() => setDockerFile(PreCheckStates.FAILED));
+  };
```

Comment:
I think you should divide this `callChecks` into 3 smaller callbacks
also prefer to enclose those functions into `useCallback`.